client/state/themes/upload-theme/reducer.js

```diff
@@ -0,0 +1,71 @@
+/**
+ * External dependencies
+ */
+import { combineReducers } from 'redux';
+import { omit } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { createReducer } from 'state/utils';
+import {
+	THEME_UPLOAD_START,
+	THEME_UPLOAD_SUCCESS,
+	THEME_UPLOAD_FAILURE,
+	THEME_UPLOAD_CLEAR,
+	THEME_UPLOAD_PROGRESS,
+} from 'state/action-types';
+
+const clearState = function( state, { siteId } ) {
+	return omit( state, siteId );
+};
+
+// Returns a func that sets state to be the specified action property,
+// or true if no property supplied
+const setState = function( actionProp ) {
+	return ( state, action ) => {
+		const { siteId } = action;
+		const newState = Object.assign( {}, state );
+		newState[ siteId ] = actionProp ? action[ actionProp ] : true;
+		return newState;
+	};
+};
+
+export const uploadedTheme = createReducer( {}, {
+	[ THEME_UPLOAD_SUCCESS ]: setState( 'theme' ),
+	[ THEME_UPLOAD_CLEAR ]: clearState,
+	[ THEME_UPLOAD_START ]: clearState,
```

Comment:
> I think my strategy'd be rather to have reducers more individually set and unset bools 

Isn't that what they are doing? For example, `inProgress` is boolean state that gets set to `true` on `start`, and set to `false` (in reality deleted for `siteId`) on `clear`, `failure`, and `success`.

> do we even need the clearThemeUpload action at all?

We need some way to get from a successful upload where we show the theme details, to the start state that will let you upload another.